### ISSUE TITLE
Provide access to the training dataset

### ISSUE THREAD

**Issue by ivannp** [microsoft/LightGBM#5191]
There are no public APIs to directly interact with the training dataset created from users data (the result of Dataset.construct).

Provide a method to export the training dataset as a numpy array.

**Comment by jameslamb** [comment]
> There are no public APIs to directly interact with the training dataset created from users data.
>
> Provide a method to export the training dataset as a numpy array.

If you are talking about the Python package specifically (I'm assuming that, since you mentioned `numpy`), this statement is not true. https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html describes many methods to interact with the constructed `Dataset`.

<img width="559" alt="image" src="https://user-images.githubusercontent.com/7608904/166150550-ff4925f2-8ed7-4c28-a00f-f9cd94e083a0.png">

I *think* you mean that you'd like something like this to work:

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

dtrain = lgb.Dataset(
    data=X,
    label=y,
    free_raw_data=True
).construct()

lgb_mat = dtrain.get_dataset_data()
```

Where a method like `get_dataset_data()` returns, in `numpy` format, the actual training data used by LightGBM (e.g. with continuous features replaced by their bin IDs, composite features created by Exclusive Feature Bundling, etc.).

Is that what you're asking for? If not, could you please provide some more specifics around what you're looking for that LightGBM doesn't currently provide?

**Comment by ivannp** [comment]
hi James, `dtrain.get_dataset_data()` from your example is exactly what I was asking for.

**Comment by StrikerRUS** [comment]
@ivannp You can use [`get_data()`](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset.get_data) method for this purpose.
```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression()

dtrain = lgb.Dataset(
    data=X,
    label=y,
    free_raw_data=False
).construct()

lgb_mat = dtrain.get_data()
```

**Comment by StrikerRUS** [comment]
Ah, nevermind, ignore my previous comment. You need 
> the **actual** training data used by LightGBM (e.g. with continuous features replaced by their bin IDs, composite features created by Exclusive Feature Bundling, etc.)

**Comment by jameslamb** [comment]
>  is exactly what I was asking for

Ok thanks! We can take this as a feature request then, I think it could be really useful for understanding more about the decisions LightGBM is making.

Per this project's policy, I'm going to close this issue and add it to the list of feature requests at #2302. Anyone wanting to contribute this can leave a comment here to ask for help.

To be honest though, the level of maintainer activity in this repo has been fairly low recently and we have some other large initiatives we're trying to get in to create a v4.0.0 release (#5153), so unless you or someone else from the community wants to work on this, I don't think this will be implemented soon.

**Comment by StrikerRUS** [comment]
I'm not exactly sure, but looks like internal Dataset method `_dump_text()` can be almost the thing that is requested in this issue.
```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression()

dtrain = lgb.Dataset(
    data=X,
    label=y,
    free_raw_data=True
).construct()

dtrain._dump_text('dataset.txt')
```
https://github.com/microsoft/LightGBM/blob/924d337de36352ea9f8442054df23db0624a1eb6/python-package/lightgbm/basic.py#L2526-L2529

Precisely, this part of the underlying cpp function:
https://github.com/microsoft/LightGBM/blob/417c732cc05978276d533c43070d24dfd6c892b7/src/io/dataset.cpp#L1120-L1138

**Comment by vladv14** [comment]
It would also be useful to return bin upper bounds of the constructed dataset, in a similar format to how you can provide them using `forced_bins` Dataset parameter.
https://github.com/microsoft/LightGBM/blob/master/examples/regression/forced_bins.json

So you can maybe have a method like `get_dataset_data()` in `Dataset`, which will return a numpy array with continuous features replaced by their bin IDs, as mentioned above, plus the bin upper bounds metadata. Or perhaps those could be split out into two separate methods like `get_data_bin_ids()`, returning a numpy array, and `get_data_bin_upper_bounds()`, returning a dictionary in the format of forced_bins.json

**Comment by jameslamb** [comment]
Sorry, this was locked accidentally. Just unlocked it. We'd still love help with this feature!

**Comment by pj-ml** [comment]
For anybody looking for a not so ideal temporary solution to displaying their lightgbm dataset as a pandas dataset (I use polars then convert to pandas, but you could probably use pandas directly too):
`import polars as pl
import os

def parse_lightgbm_dataset_text(path):
    # Split the text into lines
    with open(path, 'r') as f:
        feature_numeric_lines_found = False
        
        for idx, line in enumerate(f):    
            if line.startswith('num_total_features:'):
                num_features = int(line.split(': ')[-1])
            elif line.startswith('feature_names:'):
                feature_names = [feature for feature in line.split(': ')[-1].split(', ') if feature and feature!='\n']
            if line.startswith('feature '):
                feature_numeric_lines_found = True

            if feature_numeric_lines_found and not line.startswith('feature '):
                start_line = idx
                break

    # Read the feature data using polars
    if start_line is not None:
        df = pl.read_csv(
            path, 
            skip_rows=start_line,
            has_header=False,
            new_columns=feature_names,
            separator=",",
            null_values=["NA", " NA", " "],  # Treat 'NA' and ' NA' as missing values
            dtypes={col: pl.Int16 for col in feature_names},  # Set datatype for each column
            truncate_ragged_lines=True  # Remove trailing empty columns
        )
        # drop the last empty column
        df = df.drop(df.columns[-1])
                
        return df.to_pandas()
    return None

def lightgbm_dataset_as_pandas(lgb_dataset, tmp_path='_temp_lightgbm_dataset.txt'):
    # Get the text representation of the dataset
    lgb_dataset._dump_text(tmp_path)

    try:
        # Parse the text and return the resulting DataFrame
        result_df = parse_lightgbm_dataset_text(tmp_path)
    finally:
        # Remove the temporary file
        os.remove(tmp_path)
    
    return result_df

lightgbm_dataset_as_pandas(your_lightgbm_dataset)
`

**Comment by AxelQwant** [comment]
I have an issue loading the dataset from a binary file :
```python
dataset = lgb.Dataset("./lgb_dataset_valid.bin", free_raw_data=False)
dataset.construct()
print(dataset.feature_name)
print(dataset.label)
print(dataset.label.shape)
print(dataset.data)
```
Outputs :
```sh
[LightGBM] [Info] Load from binary file [./lgb_dataset_valid.bin](https://vscode-remote+attached-002dcontainer-002b7b22636f6e7461696e65724e616d65223a222f6d6f6e6f7265706f5f646576636f6e7461696e6572222c2273657474696e6773223a7b22636f6e74657874223a226f7262737461636b227d7d.vscode-resource.vscode-cdn.net/root/python-mono-repo/apps/rank2/notebooks/lgb_dataset_valid.bin)
['attribute(domain_rank)', 'attribute(host_rank)', 'attribute(page_rank)', 'attribute(quality_signal_malware_score)'...]
[1. 1. 0. ... 0. 1. 1.]
(156909,)
"./lgb_dataset_valid.bin"
```
Why can I access the data for the features names', the labels, but not the actual data ?
I would like to access the numpy array, not the string to the file
Is there a workaround ?

**Comment by jameslamb** [comment]
Please, @AxelQwant , re-read the posts above yours. LightGBM's `Dataset` does not save the raw input data, only the binned representation.

And the `.data` attribute on the `Dataset` in Python specifically holds whatever you passed in when you ran `lgb.Dataset()`... which might be a `numpy` array, `pandas` dataframe, filepath, etc.

You can try the workaround posted in comments above, and stay subscribed to the notifications for this feature request to be notified if the ability to dump out LightGBM's internal representation as a `numpy` array is ever formally added to the library.
